File: backend/schemas/calendar.py
from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field


class CalendarEventCreate(BaseModel):
    title: str = Field(min_length=1, max_length=160)
    description: Optional[str] = None
    event_type: str = "MEETING"
    starts_at: datetime
    ends_at: datetime
    location: Optional[str] = None
    meeting_url: Optional[str] = None
    attendee_ids: List[int] = Field(default_factory=list)
    recurrence_mode: str = "none"
    recurrence_until: Optional[datetime] = None


class CalendarEventUpdate(BaseModel):
    title: Optional[str] = Field(default=None, min_length=1, max_length=160)
    description: Optional[str] = None
    event_type: Optional[str] = None
    starts_at: Optional[datetime] = None
    ends_at: Optional[datetime] = None
    location: Optional[str] = None
    meeting_url: Optional[str] = None
    attendee_ids: Optional[List[int]] = None
    recurrence_mode: Optional[str] = None
    recurrence_until: Optional[datetime] = None


class CalendarAvailabilityCreate(BaseModel):
    user_id: Optional[int] = None
    status: str = "VACATION"
    title: Optional[str] = Field(default=None, max_length=160)
    starts_at: datetime
    ends_at: datetime
    all_day: bool = True
    note: Optional[str] = None


class CalendarAvailabilityUpdate(BaseModel):
    user_id: Optional[int] = None
    status: Optional[str] = None
    title: Optional[str] = Field(default=None, max_length=160)
    starts_at: Optional[datetime] = None
    ends_at: Optional[datetime] = None
    all_day: Optional[bool] = None
    note: Optional[str] = None


class CalendarAvailabilityOut(BaseModel):
    id: int
    project_id: int
    user_id: int
    user_name: Optional[str] = None
    user_email: Optional[str] = None
    user_avatar_url: Optional[str] = None
    created_by_id: int
    created_by_name: Optional[str] = None
    status: str
    title: Optional[str] = None
    starts_at: datetime
    ends_at: datetime
    all_day: bool = True
    note: Optional[str] = None
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class CalendarEventOut(BaseModel):
    id: int
    project_id: int
    title: str
    description: Optional[str] = None
    event_type: str
    starts_at: datetime
    ends_at: datetime
    location: Optional[str] = None
    meeting_url: Optional[str] = None
    attendee_ids: List[int] = Field(default_factory=list)
    recurrence_series_id: Optional[str] = None
    recurrence_mode: str = "none"
    recurrence_until: Optional[datetime] = None
    created_by_id: int
    created_by_name: Optional[str] = None
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True

File: frontend/services/calendar.ts
import api from "@/lib/axios";
import { dedupeRequest } from "@/lib/request-dedupe";

export type CalendarEventType =
  | "MEETING"
  | "DAILY"
  | "PLANNING"
  | "REVIEW"
  | "RETRO"
  | "FOCUS"
  | "OTHER";

export interface CalendarEvent {
  id: number;
  project_id: number;
  title: string;
  description?: string | null;
  event_type: CalendarEventType | string;
  starts_at: string;
  ends_at: string;
  location?: string | null;
  meeting_url?: string | null;
  attendee_ids: number[];
  recurrence_series_id?: string | null;
  recurrence_mode?: "none" | "daily" | "weekdays" | "weekly" | string;
  recurrence_until?: string | null;
  created_by_id: number;
  created_by_name?: string | null;
  created_at: string;
  updated_at?: string | null;
}

export interface CalendarEventCreate {
  title: string;
  description?: string | null;
  event_type: CalendarEventType | string;
  starts_at: string;
  ends_at: string;
  location?: string | null;
  meeting_url?: string | null;
  attendee_ids: number[];
  recurrence_mode?: "none" | "daily" | "weekdays" | "weekly" | string;
  recurrence_until?: string | null;
}

export type CalendarEventUpdate = Partial<CalendarEventCreate>;

export const getProjectCalendarEvents = async (
  projectId: number,
  params: { start?: string; end?: string } = {}
): Promise<CalendarEvent[]> => {
  const paramsKey = JSON.stringify(params);

  return dedupeRequest(`calendar:events:${projectId}:${paramsKey}`, async () => {
    const response = await api.get(`/calendar/project/${projectId}`, { params });
    return response.data;
  });
};

export const createProjectCalendarEvent = async (
  projectId: number,
  data: CalendarEventCreate
): Promise<CalendarEvent> => {
  const response = await api.post(`/calendar/project/${projectId}`, data);
  return response.data;
};

export const updateProjectCalendarEvent = async (
  eventId: number,
  data: CalendarEventUpdate
): Promise<CalendarEvent> => {
  const response = await api.put(`/calendar/${eventId}`, data);
  return response.data;
};

export const deleteCalendarEvent = async (
  eventId: number
): Promise<{ message: string }> => {
  const response = await api.delete(`/calendar/${eventId}`);
  return response.data;
};

export const deleteCalendarEventSeries = async (
  eventId: number
): Promise<{ message: string; deleted: number }> => {
  const response = await api.delete(`/calendar/${eventId}/series`);
  return response.data;
};


export type CalendarAvailabilityStatus =
  | "AVAILABLE"
  | "UNAVAILABLE"
  | "VACATION"
  | "SICK_LEAVE"
  | "FOCUS_TIME";

export interface CalendarAvailability {
  id: number;
  project_id: number;
  user_id: number;
  user_name?: string | null;
  user_email?: string | null;
  user_avatar_url?: string | null;
  created_by_id: number;
  created_by_name?: string | null;
  status: CalendarAvailabilityStatus | string;
  title?: string | null;
  starts_at: string;
  ends_at: string;
  all_day: boolean;
  note?: string | null;
  created_at: string;
  updated_at?: string | null;
}

export interface CalendarAvailabilityCreate {
  user_id?: number | null;
  status: CalendarAvailabilityStatus | string;
  title?: string | null;
  starts_at: string;
  ends_at: string;
  all_day?: boolean;
  note?: string | null;
}

export type CalendarAvailabilityUpdate = Partial<CalendarAvailabilityCreate>;

export const getProjectCalendarAvailability = async (
  projectId: number,
  params: { start?: string; end?: string; user_id?: number } = {}
): Promise<CalendarAvailability[]> => {
  const paramsKey = JSON.stringify(params);

  return dedupeRequest(`calendar:availability:${projectId}:${paramsKey}`, async () => {
    const response = await api.get(`/calendar/project/${projectId}/availability`, { params });
    return response.data;
  });
};

export const createProjectCalendarAvailability = async (
  projectId: number,
  data: CalendarAvailabilityCreate
): Promise<CalendarAvailability> => {
  const response = await api.post(`/calendar/project/${projectId}/availability`, data);
  return response.data;
};

export const updateProjectCalendarAvailability = async (
  availabilityId: number,
  data: CalendarAvailabilityUpdate
): Promise<CalendarAvailability> => {
  const response = await api.put(`/calendar/availability/${availabilityId}`, data);
  return response.data;
};

export const deleteProjectCalendarAvailability = async (
  availabilityId: number
): Promise<{ message: string }> => {
  const response = await api.delete(`/calendar/availability/${availabilityId}`);
  return response.data;
};
